ossfuzz/aspell_fuzzer.cpp

```diff
@@ -0,0 +1,199 @@
+
+  // Iterate over all misspellings.
+  token = aspell_document_checker_next_misspelling(doc_checker);
+
+  FUZZ_DEBUG("Token len %d", token.len);
+
+  for (;
+       token.len != 0;
+       token = aspell_document_checker_next_misspelling(doc_checker))
+  {
+    // Get spelling suggestions for the misspelling.
+    auto word_list = aspell_speller_suggest(spell_checker,
+                                            data_str + token.offset,
+                                            token.len);
+
+    // Iterate over the suggested replacement words in the word list.
+    AspellStringEnumeration *els = aspell_word_list_elements(word_list);
+
+    for (const char *word = aspell_string_enumeration_next(els);
+         word != 0;
+         word = aspell_string_enumeration_next(els))
+    {
+      // Conditionally print out the suggested replacement words.
+      FUZZ_DEBUG("Suggesting replacement for word at offset %d len %d: %s",
+                 token.offset,
+                 token.len,
+                 word);
+    }
+    delete_aspell_string_enumeration(els);
+  }
+
+EXIT_LABEL:
+
+  if (doc_checker != NULL) {
+    delete_aspell_document_checker(doc_checker);
+  }
+
+  if (spell_checker != NULL) {
+    delete_aspell_speller(spell_checker);
+  }
+
+  if (spell_config != NULL) {
+    delete_aspell_config(spell_config);
+  }
+
+  return 0;
+}
+
+// Returns -1 on error, or the number of bytes consumed from the config string
+// otherwise.
+int parse_config(AspellConfig *spell_config,
+                 uint8_t *config,
+                 size_t config_len)
+{
+  uint8_t line[MAX_CONFIG_LEN];
+
+  uint8_t *config_ptr = config;
+  size_t config_ptr_used = 0;
+
+  uint8_t *delimiter;
+
+  // Iterate over the lines.
+  for (delimiter = (uint8_t *)memchr(config_ptr,
+                                     '\n',
+                                     config_len - config_ptr_used);
+       delimiter != NULL;
+       delimiter = (uint8_t *)memchr(config_ptr,
+                                     '\n',
+                                     config_len - config_ptr_used))
+  {
+    int line_len = delimiter - config_ptr;
+
+    if (line_len == 0)
+    {
+      // The line is zero-length; it's the end of configuration. Skip over the
+      // delimiter and break out.
+      FUZZ_DEBUG("Breaking out of config");
+      config_ptr++;
+      config_ptr_used++;
+      break;
+    }
+
+    // Copy the line into the line array. Replace the newline by a null.
+    memcpy(line, config_ptr, line_len);
+    line[line_len] = 0;
+
+    // Try and split the line by =.
+    uint8_t *kv_delim = (uint8_t *)memchr(line, '=', line_len);
+
+    if (kv_delim == NULL)
+    {
+      // Can't split as a k/v pair. Exit early.
+      return -1;
+    }
+
+    // Convert the line into a key, value pair.
+    kv_delim[0] = 0;
+
+    char *keyword = reinterpret_cast<char *>(line);
+    char *value = reinterpret_cast<char *>(kv_delim + 1);
```

Comment:
It doesn't look like you are removing trailing and leading white space.  This needs to be done or otherwise `aspell_config` will reject the key and likely the value.  If it doesn't the value is likely not to return the results expected.

Reply:
As I mentioned before the fuzzer will learn the right inputs over time. Feeding it a sample corpus will certainly help. Again, having arbitrary input here is a good thing - getting rejected will terminate the test fast, and the fuzzer will learn not to do that (because it's not an interesting input).